### PR TITLE
GH-1700: Check Partition in DLPublishingRecoverer

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-kafka/issues/1700

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java [hunk 1]
```diff
@@ -195,6 +232,30 @@ public void accept(ConsumerRecord<?, ?> record, Exception exception) {
...
+			boolean anyMatch = partitions.stream().anyMatch(pi -> pi.partition() == tp.partition());
+			if (!anyMatch) {
+				this.logger.warn(() -> "Destination resolver returned non-existent partition " + tp
+						+ ", KafkaProducer will determine partition to use for this topic");
+				return new TopicPartition(tp.topic(), -1);
+			}
+			return tp;
+		}
+		catch (Exception ex) {
+			this.logger.debug(ex, () -> "Could not obtain partition info for " + tp.topic());
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java:254]: Just to double check: is it intentional to have it as `debug`  over here on this error and it is not a copy/paste artifact from the code above?
I mean returned wrong partition is a `warn`, but this exception is `debug` 🤷

File: spring-kafka/src/main/java/org/springframework/kafka/listener/RecoveryStrategy.java [hunk 2]
```diff
@@ -35,10 +36,11 @@
...
 	 * @param ex the exception.
 	 * @param container the container (or parent if a child container).
+	 * @param consumer the consumer.
 	 * @return true to skip.
 	 * @throws InterruptedException if the thread is interrupted.
 	 */
-	boolean recovered(ConsumerRecord<?, ?> record, Exception ex, @Nullable MessageListenerContainer container)
-			throws InterruptedException;
+	boolean recovered(ConsumerRecord<?, ?> record, Exception ex, @Nullable MessageListenerContainer container,
+			@Nullable Consumer<?, ?> consumer) throws InterruptedException;
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/RecoveryStrategy.java:44]: Looks like a breaking change in the public API signature...

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java (hunk 1 above)
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java:254]: It's intentional - we might not have permissions and I didn't want to flood the logs.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/RecoveryStrategy.java (hunk 2 above)
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/RecoveryStrategy.java:44]: It's not public; it's package-private.

Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/RecoveryStrategy.java:44]: Oops - sorry - no - this interface is new in 2.7.

**Review by artembilan — APPROVED** [review]

**Pull request merged**